Added Angular guide (#222)
ckeditor/ckeditor-docs: docs/guide/dev/integration/angular/README.md

export class MyComponent {
    public editorData = '<p>Hello, world!</p>';
}
```

```html
<ckeditor [data]="editorData"></ckeditor>
```

### `tagName`

Specifies the tag name of the HTML element on which the editor will be created.

The default tag is `<textArea>`.
``` html
<ckeditor tagName="textarea"></ckeditor>
```

### `readOnly`

Editor can be set to {@link guide/dev/features/readonly/README read-only mode} via `readOnly` property:

```html
<ckeditor
	data="<p>Editor's content</p>"
	readOnly="true"
><ckeditor>
```

This property takes precedence over {@linkapi CKEDITOR.config#readOnly `config.readOnly`} setting.

## Supported `@Output` properties

The following `@Output` properties are supported by CKEditor 4 Angular component:

### `ready`

Fires when the editor is ready. It corresponds with the {@linkapi CKEDITOR.editor#instanceReady `editor#instanceReady`} event.

### `change`

Fires when the content of the editor has changed. It corresponds with the {@linkapi CKEDITOR.editor#change `editor#change`} event. For performance reasons this event may be called even when data didn't really changed.

```html
<ckeditor (change)="onChange($event)"></ckeditor>
```

``` typescript

> import { CKEditor4 } from "@ckeditor/ckeditor4-angular/ckeditor";

Reviewer: We use single quotes for js strings.

@Component( {
    …
} )
export class MyComponent {
    public onChange( event: CKEditor4.EventInfo ) {
        console.log( event.editor.getData() );
    }
    …
}
```

### `dataChange`

Fires when the content of the editor has changed. In contrast to `change` - only emits when data really changed thus can be successfully used with `[data]` and two way `[(data)]` binding.

### `focus`

Fires when the editor's editable is focused. It corresponds with the {@linkapi CKEDITOR.editor#focus `editor#focus`} event.

### `blur`

Fires when the editing view of the editor is blurred. It corresponds with the {@linkapi CKEDITOR.editor.blur `editor#blur`} event.

### Note

> Callbacks bound to each of events, are called with {@linkapi CKEDITOR.eventInfo `eventInfo`}. For importing and usage of `EventInfo` interface see [**`change`** example](#change).

Reviewer: Not each event, as [`dataChange` is fired with a different signature](https://github.com/cksource/ckeditor4-angular/blob/3bfa435/src/ckeditor/ckeditor.component.ts#L293) it gets a plain string.

Reviewer: I think it should be changed in the upstream, that is `ckeditor4-angular` integration.

## Basic Usage

In order to create editor instance in Angular, install `ckeditor4-angular` npm package as a dependency of your project:

```bash
npm install --save ckeditor4-angular
```

After installing, import `CKEditorModule` to your application:

```typescript
import CKEditorModule from 'ckeditor4-angular';

@NgModule( {
	imports: [
		…
		CKEditorModule,
		…
	],
	…
} )
```

Now you can use `<ckeditor>` tag in component template to include the rich text editor:

```html
<ckeditor data="<p>Hello, world!</p>"></ckeditor>
```

The `data` attribute used in the above example is responsible for setting the editor's data.

## Customizing CKEditor Preset or Version

By default `CKEditor` Angular component loads [Standard Preset](https://ckeditor.com/docs/ckeditor4/latest/examples/standardpreset.html) of the latest CKEditor release from <a href="https://cdn.ckeditor.com/">CDN</a> while creating the first editor. This behavior can be altered by setting the value of `editorUrl` attribute in template to point to desired CKEditor location:

```html

> <ckeditor [editorUrl]="https://your-website.example/ckeditor/ckeditor.js"></ckeditor>

Reviewer: Setting `editorUrl` attribute like that will throw exception:

```
compiler.js:15859 Uncaught Error: Quotes are not supported for evaluation!
        Statement: //cdn.ckeditor.com/4.5.5/standard-all/ckeditor.js located at ng:///AppModule/AppComponent.html@3:11
```

Reviewer: Should be without brackets.

## Integration with ngModel

The component implements the [`ControlValueAccessor` interface](https://angular.io/api/forms/ControlValueAccessor) and works with `ngModel`.

To use it, first create model in your component:

``` typescript
@Component( {
    …
} )
export class MyComponent {
    public model = {
        editorData: '<p>Hello, world!</p>'
    };
    …
}
```

Then you can use the model in the template to enable a two-way binding.

```html
<ckeditor [(ngModel)]="model.editorData"></ckeditor>
```

## Supported `@Input` properties

The following `@Input` properties are supported by CKEditor 4 Angular component:

### `editorUrl`

URL address to `ckeditor.js`, for more detailed information see [Customising CKEditor Preset or Version](#customizing-ckeditor-preset-or-version).

### `config`

Custom configuration can be passed to the editor via `config` attribute passed to `<ckeditor>` in component template. The following example shows {@link guide/dev/features/toolbar/README how to change the contents of the toolbar}:

```html
<ckeditor
	data="<p>Editor' content</p>"

> 	[config]="{ { toolbar: [ [ 'Bold' ] ] } }"

Reviewer: Value of `[config]` throws parsing error to the console.

### `data`

The initial data of the editor. It can be a static value:

```html
<ckeditor data="<p>Hello, world!</p>" …></ckeditor>
```

or a shared parent component’s property

```typescript
@Component( {
    …
} )
export class MyComponent {
    public editorData = '<p>Hello, world!</p>';
    …
}
```

```html
<ckeditor [data]="editorData" …></ckeditor>
```

### `tagName`

Specifies the tag name of the HTML element on which the editor will be created.

The default tag is `<textArea>`.
``` html
<ckeditor tagName="textarea" …></ckeditor>
```

### `readOnly`

Editor can be set to {@link guide/dev/features/readonly/README read-only mode} via `readOnly` property:

```html
<ckeditor
	data="<p>Editor's content</p>"
	readOnly="true"
><ckeditor>
```

This property takes precedence over {@linkapi CKEDITOR.config#readOnly `config.readOnly`} setting.

## Supported `@Output` properties

The following `@Output` properties are supported by CKEditor 4 Angular component:

### `ready`

Fires when the editor is ready. It corresponds with the {@linkapi CKEDITOR.editor#instanceReady `editor#instanceReady`} event.

### `change`

> Fires when the content of the editor has changed. It corresponds with the {@linkapi CKEDITOR.editor#change `editor#change`} event. For performance reasons this event may be called even when data didn't really changed.

Author: Components `change` fully matches https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_editor.html#event-change the rationale comes from our docs.

---
category: integration
order: 60
url: guide/dev_angular
menu-title: Angular Integration
meta-title-short: Angular Integration
---
<!--
Copyright (c) 2003-2019, CKSource - Frederico Knabben. All rights reserved.
For licensing, see LICENSE.md.
-->

# Angular Integration

<info-box info=""> This feature is provided through the <a href="https://www.npmjs.com/package/ckeditor4-angular"><code>ckeditor4-angular</code> npm package</a>.
</info-box>

> CKEditor offers native Angular integration through `CKEditor` Angular component. It provides deep integration of CKEditor and Angular that lets you use the native features of CKEditor inside Angular component. The `CKEditor` Angular component is compatible with Angular versions 2.0+.

Reviewer: ```suggestion
CKEditor offers native Angular integration through `CKEditor` Angular component. It provides deep integration of CKEditor and Angular that lets you use the native features of CKEditor inside Angular component. The `CKEditor` Angular component is compatible with Angular versions 2.0 and later.
```

## Basic Usage

In order to create editor instance in Angular, install `ckeditor4-angular` npm package as a dependency of your project:

```bash
npm install --save ckeditor4-angular
```

After installing, import `CKEditorModule` to your application:

```typescript
import CKEditorModule from 'ckeditor4-angular';

@NgModule( {
	imports: [
	...
	CKEditorModule,
	...
	],
	...
} )
```

Now you can use `<ckeditor>` tag in component template to run the rich text editor:

```html
<ckeditor data="<p>Hello, world!</p>"></ckeditor>
```

The `data` attribute used in the above example is responsible for setting the editor's data.

## Customizing CKEditor Preset or Version

By default `CKEditor` Angular component loads standard preset of the latest CKEditor release from <a href="https://cdn.ckeditor.com/">CDN</a> while creating the first editor. This behavior can be altered by setting the value of `editorUrl` attribute in template to point to desired CKEditor location:

```html
<ckeditor [editorUrl]="https://your-website.example/ckeditor/ckeditor.js"></ckeditor>
```

> Note that attribute has to be assigned **before first component is initialized**.

Reviewer: ```suggestion
Note that attribute has to be assigned **before the first component is initialized**.
```

Alternatively you can load CKEditor before loading `CKEditor` Angular component. In that case the component will use the already loaded CKEditor:

```html
<script src="custom-ckeditor.js"></script>
<script src="main.js"></script>
```

## Choosing Editor Type

> By default `CKEditor` Angular component creates {@link guide/dev/framed/README classic editor}. To create {@link guide/dev/inline/README inline editor} `type` property with value of `inline` must be added:

Reviewer: From what I know, currently there's no support for classic editor in our angular integration. Is this section relevant to this integration?

```html
<ckeditor
	data="<p>Editor's content</p>"
	readOnly="true"
><ckeditor>
```

This property takes precedence over {@linkapi CKEDITOR.config#readOnly `config.readOnly`} setting.

## Supported `@Output` properties

The following `@Input` properties are supported by CKEditor 4 Angular component:

### `ready`

Fires when the editor is ready. It corresponds with the `editor#instanceReady` https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_editor.html#event-instanceReady event.

### `change`

Fires when the content of the editor has changed. In contrast to `change` - only emits when data really changed thus can be successfully used with `[data]` and two way `[(data)]` binding.

Fires when the content of the editor has changed. It corresponds with the `editor#change` https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_editor.html#event-change event. For performance reasons this event may be called even when data didn't really changed.

```html
<ckeditor (change)="onChange($event)"></ckeditor>
```

``` typescript

import { CKEditor4 } from "@ckeditor/ckeditor4-angular/ckeditor";

@Component( {
    ...
} )
export class MyComponent {

>     public Editor = ClassicEditor;

Reviewer: I don't understand this code, ClassicEditor is not defined.

```

After installing, import `CKEditorModule` to your application:

```typescript
import CKEditorModule from 'ckeditor4-angular';

@NgModule( {
	imports: [
	...
	CKEditorModule,
	...
	],
	...
} )
```

Now you can use `<ckeditor>` tag in component template to run the rich text editor:

```html
<ckeditor data="<p>Hello, world!</p>"></ckeditor>
```

The `data` attribute used in the above example is responsible for setting the editor's data.

## Customizing CKEditor Preset or Version

By default `CKEditor` Angular component loads standard preset of the latest CKEditor release from <a href="https://cdn.ckeditor.com/">CDN</a> while creating the first editor. This behavior can be altered by setting the value of `editorUrl` attribute in template to point to desired CKEditor location:

```html
<ckeditor [editorUrl]="https://your-website.example/ckeditor/ckeditor.js"></ckeditor>
```

Note that attribute has to be assigned **before the first component is initialized**.

Alternatively you can load CKEditor before loading `CKEditor` Angular component. In that case the component will use the already loaded CKEditor:

```html
<script src="custom-ckeditor.js"></script>

> <script src="main.js"></script>

Author: I mean bundled Angular package, as it has to be loaded after the `ckeditor.js`. What do you think, about adding a comment:

```suggestion
<script src="main.js"></script><!–– Your bundled Angular package ––>
```
or
```suggestion
<script src="your-app-bundle.js"></script>
```
or just drop this line?

> <script src="main.js"></script>

Reviewer: Both propositions look good. I think the second is cleaner 👌

### `data`

The initial data of the editor. It can be a static value:

```html
<ckeditor data="<p>Hello, world!</p>" ...></ckeditor>
```

or a shared parent component’s property

```typescript
@Component( {
    ...
} )
export class MyComponent {
    public editorData = '<p>Hello, world!</p>';
    ...
}
```

```html
<ckeditor [data]="editorData" ...></ckeditor>
```

### `tagName`

Specifies the tag name of the HTML element on which the editor will be created.

The default tag is `<textArea>`.
``` html
<ckeditor tagName="textarea" ...></ckeditor>
```

### `readOnly`

Editor can be set to {@link guide/dev/features/readonly/README read-only mode} via `readOnly` property:

```html
<ckeditor
	data="<p>Editor's content</p>"
	readOnly="true"
><ckeditor>
```

This property takes precedence over {@linkapi CKEDITOR.config#readOnly `config.readOnly`} setting.

## Supported `@Output` properties

The following `@Input` properties are supported by CKEditor 4 Angular component:

### `ready`

Fires when the editor is ready. It corresponds with the `editor#instanceReady` https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_editor.html#event-instanceReady event.

### `change`

> Fires when the content of the editor has changed. In contrast to `change` - only emits when data really changed thus can be successfully used with `[data]` and two way `[(data)]` binding.

Reviewer: "Fires when the content of the editor has changed." is duplicated **three** times, this makes it extremely redundant.

Reviewer: How `change` event can work "In contrast to `change` "? `change` is the event that is described here.

Fires when the content of the editor has changed. It corresponds with the `editor#change` https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_editor.html#event-change event. For performance reasons this event may be called even when data didn't really changed.

```html
<ckeditor (change)="onChange($event)"></ckeditor>
```

``` typescript

import { CKEditor4 } from "@ckeditor/ckeditor4-angular/ckeditor";

@Component( {
    ...
} )
export class MyComponent {
    public onChange( event: CKEditor4.EventInfo ) {
        console.log( event.editor.getData() );
    }
    ...
}
```

### `dataChange`

> Fires when the content of the editor has changed. In contrast to `change` - only emits when data really changed thus can be successfully used with `[data]` and two way `[(data)]` binding.

Reviewer: This is an exact copy of "Fires when the content of the editor has changed. In contrast to change - only emits when data really changed thus can be successfully used with [data] and two way [(data)] binding." from a previous section.

Author: I left this section as it correct and updated `change`.

### `focus`

> Fires when the editing view of the editor is focused. It corresponds with the `editor#focus` https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_editor.html#event-focus event.

Reviewer: Most common name for this is editable, let's stick to it.

```suggestion
Fires when the editor's editable is focused. It corresponds with the `editor#focus` https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_editor.html#event-focus event.
```

Same change for the blur event.

Reviewer: It should be a link, instead there's a event name followed by a full URL. Same thing for blur.

### `blur`

Fires when the editing view of the editor is blurred. It corresponds with the `editor#blur` https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_editor.html#event-blur event.

### Note

Callbacks bound to each of events, are called with {@linkapi CKEDITOR.eventInfo `eventInfo`}. For importing and usage of `EventInfo` interface see [**`change`** example](#change).

## Editor Instance

In most cases there is no need to break the encapsulation provided by `CKEditor` Angular component as editor's configuration and events handlers are configurable via component's properties. However If you need access to the {@linkapi CKEDITOR.editor } object, you can use the `instance` property of the component's instance:

```typescript
component.instance.getData();
```

Please note that this property is initialised asynchronously, when component is ready.

## Types

> CKEditor types can be installed from [@types/ckeditor](https://www.npmjs.com/package/@types/ckeditor). Note, they aren't developed by the CKEditor Team, so we can't guarantee that are up to date.

Reviewer: ```suggestion
CKEditor types can be installed from [@types/ckeditor](https://www.npmjs.com/package/@types/ckeditor). Note, they aren't maintained by the CKEditor Team, so we can't guarantee that are up to date.
```
